src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp

```diff
@@ -274,6 +327,68 @@ std::pair<bool, MergeTreeDataPartPtr> ReplicatedMergeTreePartCheckThread::findLo
     return std::make_pair(exists_in_zookeeper, part);
 }
 
+
+void ReplicatedMergeTreePartCheckThread::checkPartInZookeeper(MergeTreeDataPartPtr part, ReplicatedCheckResult & result)
+{
+    auto zookeeper = storage.getZooKeeper();
+    auto local_part_header = ReplicatedMergeTreePartHeader::fromColumnsAndChecksums(part->getColumns(), part->checksums);
+    const String part_name = part->name;
+    LOG_INFO(log, "Checking data of part {}.", part_name);
+
+    /// The double get scheme is needed to retain compatibility with very old parts that were created
+    /// before the ReplicatedMergeTreePartHeader was introduced.
+    String part_path = storage.replica_path + "/parts/" + part_name;
+    String part_znode = zookeeper->get(part_path);
+
+    try
+    {
+        ReplicatedMergeTreePartHeader zk_part_header;
+        if (!part_znode.empty())
+            zk_part_header = ReplicatedMergeTreePartHeader::fromString(part_znode);
+        else
+        {
+            String columns_znode = zookeeper->get(part_path + "/columns");
+            String checksums_znode = zookeeper->get(part_path + "/checksums");
+            zk_part_header = ReplicatedMergeTreePartHeader::fromColumnsAndChecksumsZNodes(columns_znode, checksums_znode);
+        }
+
+        if (local_part_header.getColumnsHash() != zk_part_header.getColumnsHash())
+            throw Exception(ErrorCodes::TABLE_DIFFERS_TOO_MUCH, "Columns of local part {} are different from ZooKeeper", part_name);
+
+        zk_part_header.getChecksums().checkEqual(local_part_header.getChecksums(), true);
+
+        checkDataPart(part, true, [this] { return need_stop.load(); });
+
+        if (need_stop)
+        {
+            result.status = {part_name, false, "Checking part was cancelled"};
+            result.action = ReplicatedCheckResult::Cancelled;
+            return;
+        }
+
+        LOG_INFO(log, "Part {} looks good.", part_name);
+        result.status = {part_name, true, ""};
+        result.action = ReplicatedCheckResult::DoNothing;
+        return;
+    }
+    catch (...)
+    {
+        if (isRetryableException(std::current_exception()))
+            throw;
```

Comment:
It's funny that hardware zk errors were not in the list of retriable exceptions until https://github.com/ClickHouse/ClickHouse/commit/852f397a97a8af0aecf8408cde065eb0dd99a52d
(interesting what else can be missing)